### ISSUE TITLE
[Not a bug] Music volume with VSTi/MIDIMapper

### ISSUE THREAD

**Issue by NightFright2k19** [Russian-Doom/russian-doom#307]
### Russian Doom version

Russian Doom 5.1

### What Operating System are you using?

Windows 10 x64 21H2

### What Build variant of Russian Doom are you using?

x64 (amd64, x86_64, win64)

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

I am using the [Yamaha S-YXG50 VSTi](https://veg.by/en/projects/syxg50/) with the [Coolsoft MIDIMapper](https://coolsoft.altervista.org/en/midimapper) and the [VSTDriver](https://github.com/Arakula/vstdriver/releases) as a MIDI device. In RD, music is far too loud by default with this setup. Sadly, changing music volume won't work in my case since the menu slider is not functional (nothing happens when settings are changed). 

In other ports like GZDoom or RZDoom, this issue doesn't exist. This indicates it might be something related to the sound system utilized by the port. Crispy Doom behaves like RD btw, so that's another indicator.

Is there any way this can be adjusted? Side note: Neither the MIDIMapper nor the VSTDriver have volume controls on their own and neither does Win10, so I depend on being able to change it via ingame options.

**Comment by JNechaevsky** [comment]
I'm afraid I have absolutely no idea what's going on. Current MIDI playing implementation comes directly from Chocolate Doom, where it was rewritten by @rfomin. Perhaps we could ask him for some recommendations or even move to Choco tracker?

**Comment by rfomin** [comment]
What version of Crispy Doom are you using? The current Crispy Doom 5.10.3 version uses midiproc.

**Comment by NightFright2k19** [comment]
I am using the latest in any case, Crispy (5.10.3) and Russian (latest dev version from Github autobuild). MIDI volume control works with neither of them.

**Comment by rfomin** [comment]
Do you build Crispy Doom? Crispy Doom uses a different midi player in the latest release version.

**Comment by rfomin** [comment]
Also, can you test Woof or the latest daily build of Chocolate Doom? http://latest.chocolate-doom.org/

**Comment by fabiangreffrath** [comment]
> Do you build Crispy Doom? Crispy Doom uses a different midi player in the latest release version.

Crispy 5.10.3 is the latest release and still uses midiproc. Crispy 5.11 is currently in development and uses @rfomin's native implementation, just as the dev snapshots of Choco.

**Comment by NightFright2k19** [comment]
Volume slider in latest Chocolate Doom (http://latest.chocolate-doom.org/downloads/chocolate-doom-20220204-win32.zip) works like a charm! Maybe RD can adopt its sound system, whatever they are using there right now?

**Comment by rfomin** [comment]
> Maybe RD can adopt its sound system, whatever they are using there right now?

RD already did that. I think the latest 5.1 version doesn't have midiproc already.

**Comment by NightFright2k19** [comment]
Then something is apparently wrong with it. Latest Crispy Doom also works for me.

**Comment by NightFright2k19** [comment]
Update: 
I was able to find out what was wrong. I had been using EAWPats before, and for that I had to edit the Timidity cfg path like this:
`timidity_cfg_path = "timidity.cfg"`

Apparently that's the part RD didn't like, must have handed over volume control to Timidity this way (which apparently cannot be changed, which MIGHT be the actual problem here). Anyway, after changing it back to this, volume control was functional again:
`timidity_cfg_path = ""`

So, long story short: False alert, no bug (or bug was sitting in front of the screen/keyboard, lol)! Can be closed.

**Comment by JNechaevsky** [comment]
@rfomin @fabiangreffrath thanks for elaboration! 

Well, I always said that the only orthodox synth is OPL. 🤠

To be more clear, I really have no idea what was happened, I've never used Timidity or played around with different, installable MIDI synths. @Dasperal, AFAIR you've made some changes in paths handling, so could you please advice, is it possible to do something about it to prevent such cases?

**Comment by NightFright2k19** [comment]
I was actually wondering why I had to specify timidity.cfg in the first place since the port should look for that file by default. If it's there, it is used, otherwise not. Should be simple. 

As for Timidity volume control: I think that actually works - as long as GUS patches are used.

**Comment by Dasperal** [comment]
So, this is how this music system mess works:

There are two music modules: OPL and SDL.
OPL has some native implementations and one using the SDL_Mixer library. When you select "OPL emulation" as a music device, the OPL music module is used. It initializes SDL_Mixer and it works.
For other music device options, the SDL music module is used.

When you select "GUS emulation" as a music device, RD generates Timidity config and provides it to SDL_Mixer via an environment variable. It also disables Fluidsynth via an environment variable because it has priority over Timidity in SDL_Mixer.

When you select "MIDI" as a music device, midi music by default is routed to Windows Native MIDI. On non-windows it is routed to SDL_Mixer. Non-midi music is also routed to SDL_Mixer (same for "GUS emulation").
If you have `timidity_cfg_path` set in the config file given Timidity config is provided to SDL_Mixer and midi music is routed to SDL_Mixer.
The config file also has the theoretical ability to turn on Fluidsynth in SDL_Mixer but it doesn't work. I think it's because official SDL_Mixer binaries don't have the Fluidsynth module compiled.

**Comment by rfomin** [comment]
> The config file also has the theoretical ability to turn on Fluidsynth in SDL_Mixer but it doesn't work. I think it's because official SDL_Mixer binaries don't have the Fluidsynth module compiled.

Yes, it works with SDL_Mixer from MSYS2. Fluidsynth in SDL_Mixer has chorus and reverb enabled by default which ruins a lot of MIDI tracks from PWADs.

> There are two music modules: OPL and SDL.

Chocolate Doom support only MIDI/MUS from WADs so I decided not to make more modules to minimize changes. For Woof we made four music modules: Native WIN, Fluidsynth, OPL, SDL.

**Comment by NightFright2k19** [comment]
So what would be the correct way to set up EAWPats with volume control? Via "GUS Emulation" and a separate timidity.cfg in the RD root directory? Because otherwise you'd have to choose "MIDI" as device and then you might run into the same issue again...
